rpcs3/Emu/Cell/Modules/cellPngDec.cpp

```diff
@@ -401,43 +464,43 @@ s32 pngDecOpen(ppu_thread& ppu, PHandle handle, PPStream png_stream, PSrc source
 		fmt::throw_exception("Creation of png_infop failed." HERE);
 	}
 
-	// Set a point to return to when an error occurs in libpng
-	if (setjmp(png_jmpbuf(stream->png_ptr)))
-	{
-		fmt::throw_exception("Fatal error in libpng." HERE);
-	}
-
 	// We must indicate, that we allocated more memory
 	open_info->initSpaceAllocated += sizeof(PngBuffer);
 
-	// Init the IO for either reading from a file or a buffer
 	if (source->srcSelect == CELL_PNGDEC_BUFFER)
 	{
-		// Set the data pointer and the file size
-		buffer->length = stream->source.fileSize;
+		buffer->length = stream->source.streamSize;
 		buffer->data = stream->source.streamPtr;
-
-		// Since we already read the header, we start reading from position 8
 		buffer->cursor = 8;
 	}
 
 	// Set the custom read function for decoding
-	png_set_read_fn(stream->png_ptr, buffer.get_ptr(), pngDecReadBuffer);
+	if (control_stream) 
+	{
+		if (open_param && open_param->selectChunk != 0)
+			fmt::throw_exception("Partial Decoding with selectChunk not supported yet.");
 
-	// We need to tell libpng, that we already read 8 bytes
-	png_set_sig_bytes(stream->png_ptr, 8);
+		stream->cbCtrlStream.cbCtrlStrmArg = control_stream->cbCtrlStrmArg;
+		stream->cbCtrlStream.cbCtrlStrmFunc = control_stream->cbCtrlStrmFunc;
 
-	// Read the basic info of the PNG file
-	png_read_info(stream->png_ptr, stream->info_ptr);
+		png_set_progressive_read_fn(stream->png_ptr, (void *)stream.get_ptr(), pngDecInfoCallback, pngDecRowCallback, pngDecEndCallback);
 
-	// Read the header info for future use
-	stream->info.imageWidth = png_get_image_width(stream->png_ptr, stream->info_ptr);
-	stream->info.imageHeight = png_get_image_height(stream->png_ptr, stream->info_ptr);
-	stream->info.numComponents = png_get_channels(stream->png_ptr, stream->info_ptr);
-	stream->info.colorSpace = getPngDecColourType(png_get_color_type(stream->png_ptr, stream->info_ptr));
-	stream->info.bitDepth = png_get_bit_depth(stream->png_ptr, stream->info_ptr);
-	stream->info.interlaceMethod = png_get_interlace_type(stream->png_ptr, stream->info_ptr);
-	stream->info.chunkInformation = pngDecGetChunkInformation(stream);
+		// push header tag to libpng to keep us in sync
+		try 
+		{
+			png_process_data(stream->png_ptr, stream->info_ptr, header, 8);
+		}
+		catch (...){
```

Comment:
In win32 with certain build options this a hack to catch access violations from reading/writing invalid memory and things... definitely not recommended!